### PR TITLE
Improve replica revision counter

### PR DESCRIPTION
**Pull request by shuo-wu**
#### Which issue(s) this PR fixes:

https://github.com/longhorn/longhorn/issues/8436

#### What this PR does / why we need it:

1. Remove unused replica HTTP service
2. Replace revision counter mutex lock with a go channel
3. Use a separate goroutine to increase the revision counter during write

#### Additional documentation or context

This PR implements idea 4 and 5 mentioned in [this comment](https://github.com/longhorn/longhorn/issues/6600#issuecomment-2092210119)

### CODE REVIEW

File: pkg/replica/revision_counter.go [hunk 1]
```diff
@@ -92,9 +90,30 @@ func (r *Replica) initRevisionCounter() error {
 	}
 	// Don't use r.revisionCache directly
 	// r.revisionCache is an internal cache, to avoid read from disk
-	// every time when counter needs to be updated.
-	// And it's protected by revisionLock
-	r.revisionCache = counter
+	atomic.SwapInt64(&r.revisionCache, counter)
```
Comment (ejweber) [inline pkg/replica/revision_counter.go:93]:
```suggestion
	r.revisionCache.Swap(counter) // Or similar.
```

The Golang docs suggest replacing `SwapInt64` and similar functions with `Int64`.Swap.

https://pkg.go.dev/sync/atomic#SwapInt64

This means we need to change the type of r.revisionCounter from `int64` to `atomic.Int64`, which may have other ramifications (that are hopefully positive).

The idea seems to be that the new atomic wrappers prevent developers from accidentally performing non-atomic operations on the code elsewhere.

https://medium.com/@deckarep/the-go-1-19-atomic-wrappers-and-why-to-use-them-ae14c1177ad8

IMO, in a "post Go v1.19" world, if we are switching from mutex protection to atomic protection, we should follow this convention.

Comment (ejweber) [reply pkg/replica/revision_counter.go:93]: If we make this change, other similar functions will need to change in this PR. I won't comment individually on them.

File: app/cmd/replica.go
Comment (ejweber) [inline app/cmd/replica.go:1]: Can you help clarify the need for this new context to help my understanding? I guess if we don't have it, the new revision counter goroutine will not return when a fatal issue occurs elsewhere. Does that stop the replica from terminating?

File: pkg/replica/revision_counter.go (hunk 1 above)
Comment (ejweber) [inline pkg/replica/revision_counter.go:93]: Nit. We lost the following line from the comment, but it is still relevant.

```
// every time when counter needs to be updated.
```

**Review by ejweber — COMMENTED** [review]
LGTM in principle. One significant suggestion, one nit, and one request for clarification.

File: app/cmd/replica.go
Comment (shuo-wu) [reply app/cmd/replica.go:1]: I think it's better to have a graceful way to stop such a for loop in the goroutine. Yes, theoretically, the fatal error exit should stop everything for the engine process, including this goroutine.

**Review by ejweber — APPROVED** [review]
The changes LGTM. Thanks!

**Review by PhanLe1010 — APPROVED** [review]
LGTM in gerneral. Quick question: we are removing the `/pkg/replica/rest`, are we no longer need it in any testing tool?

**Comment by shuo-wu** [comment]
> Quick question: we are removing the /pkg/replica/rest, are we no longer need it in any testing tool?

No, we don't need it anymore.

**Pull request merged**